### PR TITLE
Key pair garbage collection

### PR DESCRIPTION
**Pull request by marians**
Solves #81

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -273,6 +278,13 @@ func Initialize(configDirPath string) error {
 
 	KubeConfigPaths = getKubeconfigPaths(HomeDirPath)
 
+	// apply garbage collection
+	randSource := rand.NewSource(time.Now().UnixNano())
+	randGenerator := rand.New(randSource)
+	if randGenerator.Float32() < garbageCollectionLikelihood {
+		GarbageCollectKeyPairs()
+	}
```
Comment (xh3b4sd) [inline config/config.go]: I think we should just do it always.

**Comment by marians** [comment]
Just pulled back the review requests. Found that I still need to clean up a little.

File: config/garbage_collect.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	"os"
+	"strings"
+	"time"
+
+	"github.com/giantswarm/microerror"
+)
+
+// GarbageCollectKeyPairs removes files from expired key pairs
+func GarbageCollectKeyPairs() error {
+	files, err := ioutil.ReadDir(CertsDirPath)
```
Comment (xh3b4sd) [inline config/garbage_collect.go]: I would use a walk function even if there is no nesting at the moment.

File: config/garbage_collect_test.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+RvOTa8hYiU6A475WuZKyEHcwnGYe57u2I2KbMgcKjPniocj4QzgYsVAVKW3IwaOh
+yE+vPxsiUkvQHdO2fojCkY8jg70jxM+gu59tPDNbw3Uh/2Ij310FgTHsnGQMyA==
+-----END CERTIFICATE-----`
+
+	expired, err := isCertExpired([]byte(certPEM))
+	if err != nil {
+		t.Error(err)
+	}
+
+	if expired != true {
```
Comment (xh3b4sd) [inline config/garbage_collect_test.go]: When the method should be verified the negative case is missing.

File: config/garbage_collect_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		t.Error("Expected true, got false")
+	}
+}
+
+func TestGarbageCollectKeyPairs(t *testing.T) {
```
Comment (xh3b4sd) [inline config/garbage_collect_test.go]: We usually tend to structure test case names a bit like `Test_GarbageCollection_KeyPairs` for better readability.

File: config/garbage_collect_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	// temporary config dir
+	dir, tempConfigErr := tempConfig("")
+	if tempConfigErr != nil {
+		t.Error(tempConfigErr)
+	}
+	defer os.RemoveAll(dir)
+
+	// copy test files over to temporary certs dir
+	basePath := "testdata"
+	files, _ := ioutil.ReadDir(basePath)
```
Comment (xh3b4sd) [inline config/garbage_collect_test.go]: As I already mentioned we should aim to use abstractions for dependencies like filesystems.

**Comment by marians** [comment]
To @xh3b4sd @kopiczko: CircleCI is activated but cannot be green currently, because https://github.com/giantswarm/gsctl/pull/109 isn't merged yet.

Thanks in advance for reviewing!

**Comment by marians** [comment]
Added a test for isCertExpired for the opposite case. The other suggestions will have to wait for some indefinite time. Thanks!

**Pull request merged**